### DISCUSSION TITLE
Safety Score

### DISCUSSION THREAD

**Discussion by LelandSindt** [teslamate-org/teslamate#1997] [Ideas]
@adriankumpf 

Will you consider retrieving/recording Tesla's `Safety Score`?  (assuming it is made available via the standard API and ["published"](https://tesla-api.timdorr.com/)) 

https://www.tesla.com/support/safety-score

**Comment by aaroncohen** [comment]
Judging by the number of metrics and granularity of data that the Tesla app now provides for Safety score stuff, it may be a somewhat complicated API. It appears to offer overall score as well as scoring factor breakdown by drive and by day. It seems to also expose a fleet median for all of those values, giving you something to judge against.

**Comment by LelandSindt** [comment]
I started to look at the API endpoints today but didn't get far. 

https://github.com/timdorr/tesla-api/discussions/444

**Comment by aaroncohen** [comment]
Some initial details went in to the tesla-api repo https://github.com/timdorr/tesla-api/commit/17500b0d64dcbbe34949dddc007e4399df531a3d

**Reply by LelandSindt** [reply]
Indeed.... from this... timdorr/tesla-api#444 conversation, it sounds like they are not yet enabled...

**Reply by aaroncohen** [reply]
The insurance specific endpoints aren't yet enabled, but new Safety Score specific endpoints are. They're currently used in the latest Tesla app if you opt in through the FSD Beta sign-up button.

**Comment by corsair** [comment]
@LelandSindt gave me a poke, so posting this here and in the tesla-api repo.

Here's a complete walkthrough on how to access the Safety Score. This also comes with a new caveat that it requires the SSO Access Token and no longer utilizes the Owner's API Token (OAT). The only downside to this is that this access token must be updated every eight hours, for the request to not send a 401. If the token has expired, worry not, you simply will need to regenerate a new access token using the non-expiring SSO refresh token and you're all good to go again.

It appears that for some Tesla Accounts, Tesla changed the guid of the JWT so if you experience 403 errors (with a bunch of garbled non-UTF letters), you'll need to regenerate your refresh SSO token again.

Also important to note that ALL Owner's API endpoint API calls can also use this SSO access token, so you no longer need to keep nor generate an OAT.

To access the Safety Score metrics, you will be using the following base:
https://akamai-apigateway-vfx.tesla.com

To access the daily metrics and overall score/metrics, you'll use the following endpoint:
/safety-rating/daily-metrics

This URL also requires a handful of mandatory URL parameters which will result in an error if excluded.
vin=<vin>
timezone=<timezone> in America/Chicago format.
deviceLanguage=en
deviceCountry=US
appVersion=4.1.0-663
hasInsurance=false

**Reply by LelandSindt** [reply]
@corsair thanks for replying here.. I had initially "poked" you here but decided that it was better to tie it back to the https://github.com/timdorr/tesla-api/discussions/444 thread, so I deleted my post here.. (sorry for any confusion)

**Reply by corsair** [reply]
No worries and just saw your email as well, so hope this suffices! Provided some more documentation here as well: https://github.com/timdorr/tesla-api/issues/445

**Reply by rog0978** [reply]
For the life of me, I can't figure out what I am doing wrong...

$ curl -G -H "Authorization: Bearer _*valid SSO Access Token*_"  "https://akamai-apigateway-vfx.tesla.com/safety-rating/daily-metrics" --data-urlencode "vin=_*valid_vin*_" --data-urlencode "timezone=America/New York" --data-urlencode "deviceLanguage=en" --data-urlencode "deviceCountry=US" --data-urlencode "appVersion=4.1.1-667" --data-urlencode "hasInsurance=false"

Returns this:
{"message":"401 Unauthorized","context":{"code":401,"method":"GET","url":"http:\/\/vfx-mobile-burstapp-prod.tesla.com\/safety-rating\/daily-metrics?vin=*vin_removed*&timezone=America%2FNew%20York&deviceLanguage=en&deviceCountry=US&appVersion=4.1.1-667&hasInsurance=false","source":{"type":"NotAuthenticatedException","reason":"Unsupported input","referenceNumber":"*vin_removed*"}}}% 

I verified that the supplied token is valid (via the "original" URL where I am able to get vehicle details).

Am I missing something?

**Reply by schmug** [reply]
> @LelandSindt gave me a poke, so posting this here and in the tesla-api repo.
> 
> Here's a complete walkthrough on how to access the Safety Score. This also comes with a new caveat that it requires the SSO Access Token and no longer utilizes the Owner's API Token (OAT). The only downside to this is that this access token must be updated every eight hours, for the request to not send a 401. If the token has expired, worry not, you simply will need to regenerate a new access token using the non-expiring SSO refresh token and you're all good to go again.
> 
> It appears that for some Tesla Accounts, Tesla changed the guid of the JWT so if you experience 403 errors (with a bunch of garbled non-UTF letters), you'll need to regenerate your refresh SSO token again.
> 
> Also important to note that ALL Owner's API endpoint API calls can also use this SSO access token, so you no longer need to keep nor generate an OAT.
> 
> To access the Safety Score metrics, you will be using the following base: https://akamai-apigateway-vfx.tesla.com
> 
> To access the daily metrics and overall score/metrics, you'll use the following endpoint: /safety-rating/daily-metrics
> 
> This URL also requires a handful of mandatory URL parameters which will result in an error if excluded. vin= timezone= in America/Chicago format. deviceLanguage=en deviceCountry=US appVersion=4.1.0-663 hasInsurance=false

weird question, but what format is the VIN in?

when using the full VIN, I am getting

```json
{"message": "All of the required rules must pass for VIN",
    "context": {
        "code": 422,
        "method": "GET",
        "url": "http:\/\/vfx-mobile-burstapp-prod.tesla.com\/safety-rating\/daily-metrics",
        "validationErrors": {
            "vin": "vin must be present"}}}
```

**Reply by LelandSindt** [reply]
I am sorry, I don't recall/have not looked at this in a while.

**Comment by wmaciel** [comment]
Any updates on this? I much look forward to be able to track that in Teslamate

**Comment by rawmean** [comment]
In SafetyScore v1.2, the `rangeAggregation` record seems to be always set to null. The `dailyAggregation` is still populated. It's possible compute the weighted averages to populate `rangeAggregation`, but I am not sure if the weighs should be total miles driven or total non-AP miles driven. Any idea?
<img width="412" alt="image" src="https://user-images.githubusercontent.com/1287087/205365764-c4796ff5-2d36-4427-9d64-ca6e6b230469.png">
